conans/test/functional/compression/compression_test.py

```diff
@@ -0,0 +1,454 @@
+    compressors.append({"compressor": TarLZMACompressor()})
+
+
+# check if we can actually support symlinks:
+# some operation systems might not support them
+# e.g. Windows 10 requires either privileges or developer mode
+# also only certain file systems have support (e.g. NTFS does, but FAT32 doesn't)
+def symlinks_supported():
+    if not hasattr(os, "symlink"):
+        return False
+    client = TestClient()
+    try:
+        client.save({"a": ""})
+        with chdir(client.current_folder):
+            os.symlink("a", "b")
+            return os.path.islink("b")
+    except OSError:
+        return False
+
+
+def hardlinks_supported():
+    if not hasattr(os, "link"):
+        return False
+    client = TestClient()
+    try:
+        client.save({"a": ""})
+        with chdir(client.current_folder):
+            os.link("a", "b")
+            return os.stat("a").st_ino == os.stat("b").st_ino
+    except OSError:
+        return False
+
+
+@parameterized_class(compressors)
+@attr("slow")
+class CompressionTest(unittest.TestCase):
+    @classmethod
+    def get_file_list(cls, d):
+        files = []
+        dirs = []
+        links = []
+        for root, dirnames, filenames in os.walk(d):
+            for filename in filenames:
+                relpath = os.path.relpath(os.path.join(root, filename), d)
+                if os.path.islink(os.path.join(root, filename)):
+                    links.append(relpath)
+                else:
+                    files.append(relpath)
+            for dirname in dirnames:
+                relpath = os.path.relpath(os.path.join(root, dirname), d)
+                if os.path.islink(os.path.join(root, dirname)):
+                    links.append(relpath)
+                else:
+                    dirs.append(relpath)
+        return files, dirs, links
+
+    @classmethod
+    def is_broken_link(cls, link):
+        if os.path.islink(link):
+            target = os.readlink(link)
+            return not os.path.exists(target)
+        return False
+
+    @classmethod
+    def get_inodes(cls, d, files):
+        inodes = dict()
+        for filename in files:
+            stat = os.stat(os.path.join(d, filename))
+            if stat.st_ino not in inodes:
+                inodes[stat.st_ino] = []
+            inodes[stat.st_ino].append(filename)
+        return sorted(list(inodes.values()))
+
+    def compare_dirs(self, d1, d2):
+        files1, dirs1, links1 = self.get_file_list(d1)
+        files2, dirs2, links2 = self.get_file_list(d2)
+        self.assertEqual(files1, files2)
+        self.assertEqual(dirs1, dirs2)
+        self.assertEqual(links1, links2)
+
+        for link in links1:
+            filename1 = os.path.join(d1, link)
+            filename2 = os.path.join(d2, link)
+
+            self.assertTrue(os.path.islink(filename2))
+            target1 = os.readlink(filename1)
+            target2 = os.readlink(filename2)
+            self.assertEqual(target1, target2)
```

Comment:
I am not sure this is the test we want. If for some reason the symlink is absolute, the test might be finding the origin folder (from client1). It is totally possible to have a symlink that is ok in client1, but pointing outside the folder being packaged and it will be broken in client2. 

I think the tests should cover these cases. For that I would suggest:

- The original sources are setup, with 2 folders. One is the one being zipped, the other might contain symlinked files.
- The zipped is created from the origin folder.
- The origin folder state is captured, and the folder is completely removed, including the zipped and the other containing the symlinks targets.
- Then, the zip is decompressed in a new folder. The state is checked. Some links might be pointing to an unexisting sibling or absolute folder, and would be broken. Some other links of things contained in the zip should be ok if they were zipped with relative paths, but would be broken if absolute paths.

Reply:
I think `test_symlink_outside` covers that, wdyt?
in general, it's always the same - relative symlink before compression - you'll get relative symlink after decompression. absolute symlink before compression - you'll get absolute symlink after decompression.
in both cases, it's easy that valid symlink can become broken for client2, as well as broken client can become valid for client2 (doesn't matter relative or absolute).

Reply:
That is exactly my point. All tests pass, all test do a successful ``self.compress_decompress_check()``, that internally check that both dirs match. This shouldn't be the case, if there is a symlink pointing to something outside the thing being zipped, when unzipping in the destination it should be a broken symlink pointing to nowhere, right? I think this is what we want to test, just in case the unzipping crashes, or something like that.

Reply:
okay, there could be 4 (additional) cases:
1. absolute symlink. valid before compression, invalid after decompression.
2. absolute symlink. invalid before compression, valid after decompression.
3. relative symlink. valid before compression, invalid after decompression.
4. relative symlink. invalid before compression, valid after decompression.

I can add these tests.
please NOTE that:
a. zip/unzip will not crash in any of the cases above - it will just read/write symbolic links that are potentially invalid, but it doesn't prevent creating of them at all.
b. directories are still considered identical from my point of view - e.g. symlink point to exactly the same location for client1 and client2, but for one client destination doesn't exist.

Reply:
Yeah, exactly! And even more valid before, valid after, but being outside of the zipped. So the symlink will be able to link a file that is in the target system which might be a different one than the one in the source system (and maybe with a different absolute path, if the symlink was relative)

> a. zip/unzip will not crash in any of the cases above - it will just read/write symbolic links that are potentially invalid, but it doesn't prevent creating of them at all.

Perfect then, but better make the tests cover the cases, just in case.

> b. directories are still considered identical from my point of view - e.g. symlink point to exactly the same location for client1 and client2, but for one client destination doesn't exist.

This is what is way more problematic in my experience. If the source symlink was linking a file, but now it is a folder instead in the target system, will it link? will it crash?

Reply:
> This is what is way more problematic in my experience. If the source symlink was linking a file, but now it is a folder instead in the target system, will it link? will it crash?

it will link.
zip/tar only stores the target of the link. extracting only makes a new link pointing to the target specified. it doesn't matter if the target exists at all. and it doesn't matter what is the target for a new client, it could be:
- regular file
- pseudo-file (procfs/sysfs)
- directory
- symbolic link
- named pipe (fifo)
- socket
- device (block/character)

in any case, the symlink will be created, doesn't matter what kind of object in the file system target represents.

Reply:
> it will link.

Ok, excellent then, just wanted to make sure.

Reply:
I have added 4 additional tests for the scenarios listed above. hope it helps and it is sufficient.